File: site/js/mods/midiarp.js
/**
 * The arpeggiator on MIDI tracks.
 *
 * The browser half of `src/dnfw/mods/midiarp.py`, which carries the evidence and
 * the hardware result. Same steps, same order: check every guard and every
 * edit's stock bytes, write the edits, rebuild the N.LEN lookup from this
 * image's own length tables. `scripts/js_midiarp_check.mjs` fails if the bytes
 * differ from the Python's.
 */

import { CODE } from "./midiarp-code.js";

export const ID = "midiarp";
export const NAME = "Arpeggiator on MIDI tracks";
export const SECTION = 3;
const BASE = 0x40000400;
const LUT_BYTES = 128;

export class ModError extends Error {}

const hex = (s) => Uint8Array.from(s.match(/../g) ?? [], (b) => parseInt(b, 16));
const toHex = (a) => Array.from(a, (b) => b.toString(16).padStart(2, "0")).join("");
const i32 = (d, at) => ((d[at] << 24) | (d[at + 1] << 16) | (d[at + 2] << 8) | d[at + 3]);

export function extents() {
  return CODE.edits.map((e) => ({ section: SECTION, start: e.va - BASE,
                                  length: e.new.length / 2, what: e.what }));
}

/**
 * For each N.LEN, the trig length index nearest in duration (ties shorter);
 * INF becomes 126. The MIDI task times notes only through the trig table.
 */
export function nlenLut(content) {
  const table = (va) => Array.from({ length: 128 }, (_, k) => i32(content, va - BASE + 4 * k));
  const trig = table(CODE.trig_lengths_va).slice(0, 127);
  return Uint8Array.from(table(CODE.arp_lengths_va), (want) => {
    if (want < 0) return 126;
    let best = 0;
    for (let i = 1; i < 127; i++) {
      const d = Math.abs(trig[i] - want), bd = Math.abs(trig[best] - want);
      if (d < bd || (d === bd && trig[i] < trig[best])) best = i;
    }
    return best;
  });
}

/** -> `{ content, notes }`, the new section 3. */
export function apply(firmware) {
  const section = firmware.container.find(SECTION);
  if (section === null) throw new ModError("image has no MAIN OS section");
  const original = section.unpack();
  if (original === null) throw new ModError("MAIN OS did not depack");
  // Longer is fine: data appended after the stock end moves no address used here.
  if (original.length < CODE.stock_length) {
    throw new ModError(`MAIN OS is ${original.length.toLocaleString()} B, shorter than `
      + `${CODE.stock_length.toLocaleString()}: not Digitone II 1.11`);
  }
  const at = (va, n) => toHex(original.subarray(va - BASE, va - BASE + n));
  for (const g of CODE.guards) {
    if (at(g.va, g.bytes.length / 2) !== g.bytes) {
      throw new ModError(`0x${g.va.toString(16)} is not stock; this mod is for unmodified Digitone II 1.11`);
    }
  }
  for (const e of CODE.edits) {
    if (at(e.va, e.stock.length / 2) !== e.stock) {
      throw new ModError(`0x${e.va.toString(16)} is not stock; this mod is for unmodified `
        + "Digitone II 1.11, or another mod already wrote there");
    }
  }

  const content = original.slice();
  for (const e of CODE.edits) content.set(hex(e.new), e.va - BASE);
  content.set(nlenLut(original), CODE.lut_va - BASE);

  return {
    content,
    notes: ["ARPEGGIATOR menu opens on MIDI tracks",
            "an arp-enabled MIDI track plays its arp out over MIDI",
            `${CODE.edits.length} edits in section 3, nothing appended`],
  };
}

File: src/dnfw/mods/midiarp.py
"""The arpeggiator on MIDI tracks: a MIDI track with the arp on arpeggiates over MIDI.

Stock Digitone II firmware hides the ARPEGGIATOR menu on MIDI tracks, and the
arp itself lives only on the synth voice path, so a MIDI track's notes never
meet it. This mod opens the menu, sends an arp-enabled MIDI track's notes
through the ISR's arpeggiator, and turns each note it plays into a MIDI record
on the batch the ISR already hands the MIDI task. **Passed on the instrument on
2026-09-17** as `arp-midi-play4`: offsets, velocity, SPD, RNG and N.LEN (1/32 to
1/4 within a millisecond), no stuck notes. `docs/ideas-backlog.md` §10 carries the
whole history, and `scripts/build_arp_midi_play.py` the assembly and evidence.

## How it applies

The code is assembled ahead of time (`scripts/gen_midiarp_code.py` ->
`midiarp_code.json`), so the browser applies exactly these bytes:

1. every guard (whole instructions and the code the hooks rely on) and every
   edit's stock bytes are checked -- an image that differs is refused;
2. the edits are written: two menu bytes, five hooks, and the code cave;
3. the cave's N.LEN lookup is rebuilt from this image's own length tables.

It writes only inside section 3, changes no length, and appends nothing, so it
does not compete for the appended area or the startup hook.
"""

from __future__ import annotations

import json
import pathlib
import struct

from . import Extent, ModError, Result

ID = "midiarp"
NAME = "Arpeggiator on MIDI tracks"
SUMMARY = "The arpeggiator on MIDI tracks: the menu opens, and the arp plays out over MIDI."
DEVICE = 0x15                      # Digitone II
SECTION = 3                        # MAIN OS
BASE = 0x40000400
SPEC = json.loads((pathlib.Path(__file__).with_name("midiarp_code.json")).read_text())
LUT_BYTES = 128


def extents(firmware=None) -> list[Extent]:
    return [Extent(SECTION, e["va"] - BASE, len(e["new"]) // 2, e["what"]) for e in SPEC["edits"]]


def nlen_lut(content: bytes) -> bytes:
    """For each N.LEN, the trig length index nearest in duration (ties shorter);
    INF becomes 126, the longest finite trig length. The MIDI task times notes
    only through the trig table."""
    def table(va: int) -> list[int]:
        at = va - BASE
        return list(struct.unpack(">128i", content[at:at + 512]))
    trig = table(SPEC["trig_lengths_va"])[:127]
    return bytes(126 if want < 0 else
                 min(range(127), key=lambda i: (abs(trig[i] - want), trig[i]))
                 for want in table(SPEC["arp_lengths_va"]))


def apply(firmware) -> Result:
    section = firmware.container.find(SECTION)
    if section is None:
        raise ModError("image has no MAIN OS section")
    original = section.unpack()
    # Longer is fine: data appended after the stock end (lfowaves, bootscreen)
    # moves no address this mod writes or reads. The guards below are what
    # identify the build.
    if len(original) < SPEC["stock_length"]:
        raise ModError(f"MAIN OS is {len(original):,} B, shorter than "
                       f"{SPEC['stock_length']:,}: not Digitone II 1.11")
    for g in SPEC["guards"]:
        want = bytes.fromhex(g["bytes"])
        if original[g["va"] - BASE:g["va"] - BASE + len(want)] != want:
            raise ModError(f"0x{g['va']:08x} is not stock; this mod is for unmodified "
                           "Digitone II 1.11")
    for e in SPEC["edits"]:
        want = bytes.fromhex(e["stock"])
        if original[e["va"] - BASE:e["va"] - BASE + len(want)] != want:
            raise ModError(f"0x{e['va']:08x} is not stock; this mod is for unmodified "
                           "Digitone II 1.11, or another mod already wrote there")

    content = bytearray(original)
    for e in SPEC["edits"]:
        new = bytes.fromhex(e["new"])
        content[e["va"] - BASE:e["va"] - BASE + len(new)] = new
    lut = SPEC["lut_va"] - BASE
    content[lut:lut + LUT_BYTES] = nlen_lut(original)

    return Result(payloads={SECTION: bytes(content)}, extents=extents(),
                  notes=["ARPEGGIATOR menu opens on MIDI tracks",
                         "an arp-enabled MIDI track plays its arp out over MIDI",
                         f"{len(SPEC['edits'])} edits in section 3, nothing appended"])
